### PR TITLE
Fix float8_e4m3fn in modeling_utils

### LINKED ISSUE #32185: AttributeError: module 'torch' has no attribute 'float8_e4m3fn'
### System Info

AutoModelForCausalLM.from_pretrained now returns the above error.

### Who can help?

_No response_

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [X] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

pip install transformers

from transformers import AutoModelForCausalLM
AutoModelForCausalLM.from_pretrained(
            model_path,
            cache_dir= cache_dir, 
            device_map="auto",
            quantization_config=bnb_config,
        )

### Expected behavior

Expected to load the model.

### PR DESCRIPTION
**Pull request by SunMarc**
# What does this PR do ?

This PR fixes a bug that was introduced in the latest release. We need to check that torch.float8_e4m3fn exist (only after torch 2.1)
Fixes https://github.com/huggingface/transformers/issues/32185

### CODE REVIEW

File: src/transformers/modeling_utils.py [hunk 1]
```diff
@@ -866,9 +866,11 @@ def _load_state_dict_into_meta_model(
         module_name = param_name
         set_module_kwargs = {}
 
+        is_param_float8_e4m3fn = hasattr(torch, "float8_e4m3fn") and param.dtype == torch.float8_e4m3fn
```
Comment (ArthurZucker) [inline src/transformers/modeling_utils.py:869]: let's do this outside the for loop it will slow things down

Comment (SunMarc) [reply src/transformers/modeling_utils.py:869]: you mean the hasattr(torch, "float8_e4m3fn") check ? bc i need to check the dtype of the param afterwards

Comment (SunMarc) [reply src/transformers/modeling_utils.py]: I changed a bit, lmk if this is better

**Review by ArthurZucker — APPROVED** [review]
LGTM

**Comment by tanghengjian** [comment]
local file changed according to https://github.com/huggingface/transformers/pull/32193/files  
but the error still exists
```
  File "/opt/conda/lib/python3.10/site-packages/safetensors/torch.py", line 310, in load_file
    result[k] = f.get_tensor(k)
AttributeError: module 'torch' has no attribute 'float8_e4m3fn'
```
so , i wonder can  #32193 solve above torch error??

**Comment by amyeroberts** [comment]
Hi @tanghengjian, can you verify which version of transformers you're running in your python env? 

`python -c "import transformers; print(transformers.__version__)"`

**Pull request merged**